src/call.rs

```diff
@@ -1,72 +1,80 @@
+use crate::{Client, FromMap, TwilioError, POST};
+use serde::Deserialize;
+use std::collections::BTreeMap;
+
 pub struct OutboundCall<'a> {
-    pub from : &'a str,
-    pub to   : &'a str,
-    pub url : &'a str,
+    pub from: &'a str,
+    pub to: &'a str,
+    pub url: &'a str,
 }
 
 impl<'a> OutboundCall<'a> {
-    pub fn new(from: &'a str,to: &'a str, url: &'a str) -> OutboundCall<'a> {
-        OutboundCall { from: from, to: to, url: url }
+    pub fn new(from: &'a str, to: &'a str, url: &'a str) -> OutboundCall<'a> {
+        OutboundCall { from, to, url }
     }
 }
 
 #[derive(Debug, Deserialize)]
-#[allow(non_camel_case_types)]
+#[serde(rename_all = "lowercase")]
 pub enum CallStatus {
-    queued,
-    ringing,
-    inprogress,
-    canceled,
-    completed,
-    failed,
-    busy,
-    noanswer,
+    Queued,
+    Ringing,
+    InProgress,
+    Canceled,
+    Completed,
+    Failed,
+    Busy,
+    NoAnswer,
 }
 
 #[derive(Debug, Deserialize)]
 pub struct Call {
-    from : String,
-    to   : String,
-    sid  : String,
-    status : CallStatus,
+    from: String,
+    to: String,
+    sid: String,
+    status: CallStatus,
 }
 
-impl ::Client {
-    pub fn make_call(&self, call: OutboundCall) -> Result<Call,::TwilioError> {
-        let opts = [("To",&*call.to),("From",&*call.from),("Url",&*call.url)];
-        self.send_request(::Post,"Calls",&opts)
+impl Client {
+    pub async fn make_call(&self, call: OutboundCall<'_>) -> Result<Call, TwilioError> {
+        let opts = [
+            ("To", &*call.to),
+            ("From", &*call.from),
+            ("Url", &*call.url),
+        ];
+        self.send_request(POST, "Calls", &opts).await
     }
 }
 
-impl ::FromMap for Call {
-    fn from_map(m: &::std::collections::HashMap<&str,&str>) -> Result<Box<Call>,::TwilioError> {
-        let from = match m.get("From"){
-            Some(&v) => v,
-            None => return Err(::TwilioError::ParsingError),
+impl FromMap for Call {
+    fn from_map(mut m: BTreeMap<String, String>) -> Result<Box<Call>, TwilioError> {
```

Comment:
Curious about this change - is `BTreeMap` now considered the better choice in general than `HashMap`?

Reply:
I tried to explain the reasoning behind the type signature change in the commit message for 504df24, but the gist is, essentially as follows: the current code generates the map in either a one or two step process, depending on whether the method is a `GET` or `POST`.

If the method is a `GET`, a `Vec<(String, String)>` is created and then immediately recollected into a `HashMap<&str, &str>`. If the method is a `POST`, a `Vec<(String, String)>` is created, sorted with `Ord::cmp`, and then recollected into a `HashMap<&str, &str>`. The resulting `HashMap` is then passed by reference to `FromMap::from_map`, where nearly every single implementation then takes the contained `&str` values and calls `.to_string()` on them again, even though the original `String` values the `str` slices reference are still available. This is a lot of unnecessary work and results in quite a few redundant memory allocations.

By changing the data type to a `BTreeMap<String, String>`, passing it by-move, and eliminating the intermediate `Vec<(String, String)>`, we can eliminate some intermediate memory allocations and extra work.

1. The data structure returned by `args_from_urlencoded()` is already in a map-like form, meaning it's suitable for passing into `FromMap::from_map` and used directly. By collecting into a `BTreeMap` from the beginning, there is no need for a three-step `Vec<(String, String)> -> sort? -> HashMap<&str, &str>` conversion.
2. The map is always sorted by key according to `Ord`, meaning there's no need for an explicit `Vec::sort_by` call.
3. `FromMap::from_map` takes ownership of the map itself, and the keys and values are owned `String`s instead of `&str`, meaning the implementers of the trait can simply `remove()` owned elements from the map instead of calling `.get() -> .to_string()`.

Hope this makes some sense. Let me know if there's something I'm missing!

Reply:
Thanks for taking the time to explain it - makes sense to me.